<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [3]:
try: 
    connection = psycopg2.connect(
        dbname=DBNAME,
        user=USER,
        host=HOST,
        password=PASSWORD,
        port=PORT
    )
    print('Database connected successfully')
except:
    print('Database not connected successfully')

Database connected successfully


# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''select
COUNT(id)
FROM vacancies
'''

In [5]:
# результат запроса
vacancies_number = pd.read_sql_query(query_3_1, connection)
vacancies_number

C:\Users\User\AppData\Local\Temp\ipykernel_20608\3035423760.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_number = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''select DISTINCT
COUNT(id)
FROM employers
''' 

In [7]:
# результат запроса
employers_number = pd.read_sql_query(query_3_2, connection)
employers_number

C:\Users\User\AppData\Local\Temp\ipykernel_20608\4012475093.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_number = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''select
COUNT(id)
FROM areas
'''

In [9]:
# результат запроса
area_number = pd.read_sql_query(query_3_3, connection)
area_number

C:\Users\User\AppData\Local\Temp\ipykernel_20608\3140511850.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  area_number = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''SELECT 
COUNT(id) 
FROM industries
'''

In [11]:
# результат запроса
industires_number = pd.read_sql_query(query_3_4, connection)
industires_number

C:\Users\User\AppData\Local\Temp\ipykernel_20608\3493310793.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  industires_number = pd.read_sql_query(query_3_4, connection)


count
0    294

***

In [12]:
# выводы по предварительному анализу данных

* База данных состит из 5 таблиц;
* База данных содержит сведения о 49197 вакансиях;
* Вакансии представлены 23501 работодателем в 294 сферах деятельности;
* Вакансии распределены в 1362 регионах.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [13]:
# текст запроса
query_4_1 = f'''SELECT
COUNT(v.id) cnt,
a.name area
FROM vacancies v 
JOIN areas a ON v.area_id = a.id
GROUP BY a.name
ORDER BY 1 DESC
LIMIT 5
'''


In [14]:
# результат запроса
region_vacancies = pd.read_sql_query(query_4_1, connection)
region_vacancies

C:\Users\User\AppData\Local\Temp\ipykernel_20608\2320671827.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  region_vacancies = pd.read_sql_query(query_4_1, connection)


cnt             area
0  5333           Москва
1  2851  Санкт-Петербург
2  2112            Минск
3  2006      Новосибирск
4  1892           Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [15]:
# текст запроса
query_4_2 = '''SELECT
COUNT (id)
FROM vacancies
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
'''

In [16]:
# результат запроса
salary_exists = pd.read_sql_query(query_4_2, connection)
salary_exists

C:\Users\User\AppData\Local\Temp\ipykernel_20608\233959133.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salary_exists = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [17]:
# текст запроса
query_4_3 = '''SELECT 
AVG(salary_from) avg_min, 
AVG(salary_to) avg_max
FROM vacancies 
'''

In [18]:
# результат запроса
avg_salary_boundaries = round(pd.read_sql_query(query_4_3, connection))
avg_salary_boundaries

C:\Users\User\AppData\Local\Temp\ipykernel_20608\1667583421.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_salary_boundaries = round(pd.read_sql_query(query_4_3, connection))


avg_min   avg_max
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [19]:
# текст запроса
query_4_4 = '''SELECT
schedule schedule_type, 
employment employment_type, 
COUNT(id) cnt
FROM vacancies
GROUP BY 1, 2 
ORDER BY cnt DESC
'''

In [20]:
# результат запроса
vacancies_by_employment = pd.read_sql_query(query_4_4, connection)
vacancies_by_employment

C:\Users\User\AppData\Local\Temp\ipykernel_20608\229912.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_by_employment = pd.read_sql_query(query_4_4, connection)


schedule_type      employment_type    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [21]:
# текст запроса
query_4_5 = '''SELECT 
experience experience, 
COUNT(id) cnt
FROM vacancies 
GROUP BY experience
ORDER BY cnt 
'''

In [22]:
# результат запроса
vacancies_by_experience = pd.read_sql_query(query_4_5, connection)
vacancies_by_experience

C:\Users\User\AppData\Local\Temp\ipykernel_20608\3537855102.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_by_experience = pd.read_sql_query(query_4_5, connection)


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

In [23]:
# выводы по детальному анализу вакансий

Детальный анализ вакансий показал:
* Наибольшее количество вакансий представлено по следующим городам: Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. Два города из данного списка представляют собой зарубежные локации; 
* Верхняя или нижняя граница вилки заработной платы указана меньше, чем в половине вакансий (49%); 
* Нижняя граница вилки заработной платы составляет в среднем 71 тыс. рублей; 
* Верхняя граница вилки заработноай платы составляет в среднем 110,5 тыс. рублей; 
* Подавляющее большинство вакансий (72%) предусматривают трудоустройство с полной занятостью на полный рабочий день; 
* Большая часть вакансий (53%) предусматривают наличие опыта работы от 1 до 3 лет. Вакансии, ориентированные на более опытных кандидатов (более 6 лет опыта работы) составляют менее 3%.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [24]:
# текст запроса
query_5_1 = '''SELECT 
e.name employer,
COUNT (v.id) vacancies_cnt
FROM 
employers e JOIN vacancies v ON e.id = v.employer_id
GROUP BY employer
ORDER BY vacancies_cnt DESC
LIMIT 5
'''

In [25]:
# результат запроса
employers_leaderboard = pd.read_sql_query(query_5_1, connection)
employers_leaderboard

C:\Users\User\AppData\Local\Temp\ipykernel_20608\3714917959.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_leaderboard = pd.read_sql_query(query_5_1, connection)


employer  vacancies_cnt
0         Яндекс           1933
1     Ростелеком            491
2       Тинькофф            444
3           СБЕР            428
4  Газпром нефть            331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [26]:
# текст запроса
query_5_2 = f'''SELECT
a.name AS region,
COUNT(e.id) AS employers_cnt, 
COUNT(v.id) AS vacancies_cnt
FROM 
areas a LEFT JOIN employers e ON a.id = e.area LEFT JOIN vacancies v ON a.id = v.area_id
GROUP BY 1
HAVING COUNT (v.id) = 0
ORDER BY employers_cnt DESC
'''

In [27]:
# результат запроса
regions_employers_vacancies = pd.read_sql_query(query_5_2, connection)
regions_employers_vacancies

C:\Users\User\AppData\Local\Temp\ipykernel_20608\417205170.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  regions_employers_vacancies = pd.read_sql_query(query_5_2, connection)


region  employers_cnt  vacancies_cnt
0                          Россия            410              0
1                       Казахстан            207              0
2              Московская область             75              0
3              Краснодарский край             19              0
4              Ростовская область             18              0
..                            ...            ...            ...
588                      Струнино              0              0
589    Славгород (Алтайский край)              0              0
590                     Сорочинск              0              0
591  Ленинский (Тульская область)              0              0
592                         Тавда              0              0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [28]:
# текст запроса
query_5_3 = '''SELECT
e.name employer, 
COUNT (DISTINCT v.area_id) area_cnt
FROM 
employers e JOIN vacancies v ON e.id = v.employer_id 
GROUP BY employer
ORDER BY area_cnt DESC
'''

In [29]:
# результат запроса
employers_areas = pd.read_sql_query(query_5_3, connection)
employers_areas

C:\Users\User\AppData\Local\Temp\ipykernel_20608\2763555612.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_areas = pd.read_sql_query(query_5_3, connection)


employer  area_cnt
0                      Яндекс       181
1                  Ростелеком       152
2                  Спецремонт       116
3      Поляков Денис Иванович        88
4                   ООО ЕФИН         71
...                       ...       ...
14761                  UniSol         1
14762            UNISTORY LLC         1
14763                   UNIT6         1
14764     United Distribution         1
14765              UNITEL LLC         1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [30]:
# текст запроса
query_5_4 = '''SELECT
COUNT(e.id) employers_cnt
FROM 
employers e LEFT JOIN employers_industries ei ON e.id = ei.employer_id
WHERE ei.industry_id IS NULL 
'''

In [31]:
# результат запроса
unknown_industry = pd.read_sql_query(query_5_4, connection)
unknown_industry

C:\Users\User\AppData\Local\Temp\ipykernel_20608\3112314098.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  unknown_industry = pd.read_sql_query(query_5_4, connection)


employers_cnt
0           8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [32]:
# текст запроса
query_5_5 = '''SELECT
e.name AS employer,
COUNT(ei.industry_id) AS industry_cnt
FROM 
employers e JOIN employers_industries ei ON e.id = ei.employer_id
GROUP BY employer
HAVING COUNT (ei.industry_id) = 4
ORDER BY employer  
LIMIT 3
'''

In [33]:
# результат запроса
employers_industries = pd.read_sql_query(query_5_5, connection)
employers_industries

C:\Users\User\AppData\Local\Temp\ipykernel_20608\484412620.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_industries = pd.read_sql_query(query_5_5, connection)


employer  industry_cnt
0  101 Интернет             4
1      21vek.by             4
2          2ГИС             4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [34]:
# текст запроса
query_5_6 = '''SELECT
COUNT(ei.employer_id)
FROM employers_industries ei JOIN industries i ON ei.industry_id = i.id
WHERE i.name = 'Разработка программного обеспечения'

'''

In [35]:
# результат запроса
developers_num = pd.read_sql_query(query_5_6, connection)
developers_num

C:\Users\User\AppData\Local\Temp\ipykernel_20608\1991991405.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  developers_num = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [36]:
# код для получения списка городов-милионников
import requests # Импортируем библиотеку requests
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' # Определяем значение URL страницы для запроса
response = requests.get(url) # Делаем GET-запрос к ресурсу и результат ответа сохраняем в переменной response
print(response)# Выводим значение response на экран как объект

<Response [403]>


In [37]:
#Используем ссылку на другой ресурс, т.к. запросы на Википедию блокируются бот-фильтром 
import requests
from bs4 import BeautifulSoup

# URL статьи Википедии о городах-миллионниках
url = "https://ru.ruwiki.ru/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8"

# Отправляем запрос
headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
}
response = requests.get(url, headers=headers)

if response.status_code == 200:
    soup = BeautifulSoup(response.content, "html.parser")
    
    # Находим таблицу городов-миллионников
    table = soup.find("table", class_="wikitable")
    if table:
        cities_million = []
        # Проходим по строкам таблицы (пропускаем заголовок)
        for row in table.find_all("tr")[1:]:
            cols = row.find_all("td")
            if len(cols) >= 2:
                city = cols[0].get_text(strip=True)
                # Удаляем ссылки и примечания в скобках
                city = city.split("[")[0].strip()
                cities_million.append(city)
        
    else:
        print("Таблица не найдена. Возможно, изменилась структура страницы.")
else:
    print(f"Ошибка загрузки страницы: {response.status_code}")

In [38]:
# текст запроса
# преобразуем список городов-миллионников в строку для передачи в sql-запрос
cities_million_str = "('" + "', '".join([str(item) for item in cities_million]) + "')"
query_5_7 = f'''
SELECT
    a.name, 
    COUNT(v.id) cnt
FROM 
    areas a JOIN vacancies v ON a.id = v.area_id JOIN employers e ON v.employer_id = e.id 
WHERE 
    e.name = 'Яндекс' AND a.name IN {cities_million_str}
GROUP BY a.name
UNION ALL
SELECT 
    'total', 
    COUNT(v.id)
FROM areas a JOIN vacancies v ON a.id = v.area_id JOIN employers e ON v.employer_id = e.id 
WHERE 
    e.name = 'Яндекс' AND a.name IN {cities_million_str}
ORDER BY 2    
'''


In [39]:
# результат запроса
yandex_vacancies = pd.read_sql_query(query_5_7, connection, params=[cities_million_str])
yandex_vacancies


C:\Users\User\AppData\Local\Temp\ipykernel_20608\912994898.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  yandex_vacancies = pd.read_sql_query(query_5_7, connection, params=[cities_million_str])


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            total  485

***

In [40]:
# выводы по анализу работодателей

На основании анализа данных по работодателям можно сделать следующие выводы: 
* Наибольшее число вакансий представлено в секторах ИТ, телеком, банковская сфера, нефтяная отрасль;
* Самое большое количество работодателей зарегистрировано в Москве, Санкт-Петербурге, Минске, Алматы и Новосибирске;
* В этих же локациях представлено самое большое число вакансий;
* В части широты присутствия вакансий в регионах лидерами являются крупные федеральные компании Яндекс и Ростелеком, а также менее известные работодатели: Спецремонт (ремонт бытовой техники), Поляков Д.И., ООО ЕФИН (разработка ПО);
* У 8,4 тыс. работодателей (~ 36%) не указана сфера деятельности;
* Лидерами по наличию вакансий в различных сферах деятельности являются компании Модуль (16 сфер), Простые решения (11 сфер), Энергия (10 сфер);
* В сфере разработки ПО работают 3,5 тыс. компаний.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [41]:
# текст запроса
query_6_1 = f'''
SELECT 
    name, 
    COUNT(id)
FROM 
    vacancies
WHERE 
    LOWER(name) LIKE '%data%'
    OR LOWER(name) LIKE '%данн%'
GROUP BY
    name
UNION ALL
SELECT
    'Total',
    COUNT(id) cnt
FROM
    vacancies
WHERE
    LOWER(name) LIKE '%data%'
    OR LOWER(name) LIKE '%данн%'
ORDER BY 2 DESC
'''

In [42]:
# результат запроса
data_vacancies = pd.read_sql_query(query_6_1, connection)
data_vacancies

C:\Users\User\AppData\Local\Temp\ipykernel_20608\2935617630.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data_vacancies = pd.read_sql_query(query_6_1, connection)


name  count
0                                        Total   1771
1                              Аналитик данных     78
2                                Data Engineer     63
3                               Data Scientist     55
4                                 Data Analyst     51
..                                         ...    ...
973             Data-разработчик/Data Engineer      1
974               DBA Администратор баз данных      1
975    DBA администратор баз данных PostgreSQL      1
976  DBA / Администратор баз данных PostgreSQL      1
977                   DevOps Big Data platform      1

[978 rows x 2 columns]

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [43]:
# текст запроса
query_6_2 = f'''
SELECT
    COUNT(*)
FROM
    vacancies
WHERE
    (
        LOWER(name) ILIKE '%junior%'
        OR LOWER(experience) ILIKE '%нет опыта%'
        OR LOWER(employment) ILIKE '%стажировка%'
    )
    AND
    (
        LOWER(name) ILIKE '%data scientist%'
        OR LOWER(name) ILIKE '%data science%'
        OR LOWER(name) ILIKE '%исследователь данных%'
        OR (name ILIKE '%ML%' AND name NOT ILIKE '%html%') 
        OR LOWER(name) ILIKE '%machine learning%'
        OR LOWER(name) ILIKE '%машинн%обучен%'
    )
'''

In [44]:
# результат запроса
total_jun_ds_vacancies = pd.read_sql_query(query_6_2, connection)
total_jun_ds_vacancies

C:\Users\User\AppData\Local\Temp\ipykernel_20608\3890558577.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  total_jun_ds_vacancies = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [45]:
# текст запроса
query_6_3 = f'''
SELECT 
    COUNT(*)
FROM 
    vacancies 
WHERE 
    (    
        name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name ILIKE '%ML%' AND name NOT ILIKE '%html%') 
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%'
    )
    AND 
    (
        key_skills ILIKE '%postgres%'
        OR key_skills ILIKE '%SQL%'        
    )     
'''

In [46]:
# результат запроса
sql_requrements = pd.read_sql_query(query_6_3, connection)
sql_requrements

C:\Users\User\AppData\Local\Temp\ipykernel_20608\443994979.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  sql_requrements = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [47]:
# текст запроса
query_6_4 = f'''
SELECT 
    COUNT(*)
FROM 
    vacancies 
WHERE 
    (    
        name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name ILIKE '%ML%' AND name NOT ILIKE '%html%') 
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%'
    )
    AND 
        key_skills ILIKE '%python%'
'''

In [48]:
# результат запроса
python_skills = pd.read_sql_query(query_6_4, connection)
python_skills

C:\Users\User\AppData\Local\Temp\ipykernel_20608\950132531.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  python_skills = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [49]:
# текст запроса
query_6_5 = f''' 
SELECT
    AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1) AS skill_cnt 
FROM 
    vacancies
WHERE 
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name LIKE '%ML%' AND name NOT ILIKE '%html%') 
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%'
'''

In [50]:
# результат запроса
avg_skills = round(pd.read_sql_query(query_6_5, connection), 2)
avg_skills

C:\Users\User\AppData\Local\Temp\ipykernel_20608\2475840040.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_skills = round(pd.read_sql_query(query_6_5, connection), 2)


skill_cnt
0       6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [51]:
query_6_6 = f''' 
SELECT
    AVG((COALESCE(salary_from, salary_to) + COALESCE(salary_to, salary_from)) / 2)
FROM 
    vacancies
WHERE 
    (
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name LIKE '%ML%' AND name NOT ILIKE '%html%') 
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%'
    )
AND
    (
    salary_from > 0  
    OR salary_to > 0
    )
AND 
    experience = 'От 3 до 6 лет'
''' 

In [52]:
# результат запроса
avg_ds_salary = round(pd.read_sql_query(query_6_6, connection))
avg_ds_salary

C:\Users\User\AppData\Local\Temp\ipykernel_20608\378204387.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_ds_salary = round(pd.read_sql_query(query_6_6, connection))


avg
0  243115.0

In [53]:
# выводы по предметному анализу

На основании предметного анализа данных о вакансиях в сфере DS можно сделать следующие выводы: 
* Данные содержат сведения о 1771 вакансии в сфере Data Science (~ 3,6% от общего числа вакансий); 
* Для начинающих специалистов в Data Science есть ~ 50 вакансий, что составляет менее 3% от общего числа вакансий в этой области. Таким образом, наличие опыта работы является важным фактором для успешного трудоустройства; 
* Владение SQL в качестве одного из ключевых навыков указано в ~ 13% вакансий; 
* Владение Python в перечне ключевых навыков указано в ~ 20% вакансий. Вероятно, фактическая востребованность данных навыков в Data Science выше, т.к. часть вакансий содержит общие описания направлений деятельности (например, «анализ данных», «работа с базами данных» и т.п.); 
* Описание вакансий в сфере Data Science содержит в среднем ~ 6,5 ключевых навыков; 
* Средняя заработная плата по всем вакансиям в Data Science составляет ~ 177 тыс. рублей. Для специалистов с опытом работы от 1 года до 3 лет этот показатель составляет ~ 140 тыс. рублей, вакансии для опытных кандидатов (от 3 до 6 лет) предусматривают среднее вознаграждение 243 тыс. рублей. Таким образом очевидна зависимость уровня оплаты труда от опыта работы.


# Общий вывод по проекту

In [54]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

In [55]:
# проведем анализ средней з\п по всем данным 
query_7_1 = f''' 
SELECT
    AVG((COALESCE(salary_from, salary_to) + COALESCE(salary_to, salary_from)) / 2)
FROM 
    vacancies
WHERE 
    salary_from > 0  
    OR salary_to > 0
''' 

In [56]:
# средняя заработная плата по всем вакансиям
avg_salary = round(pd.read_sql_query(query_7_1, connection))
avg_salary

C:\Users\User\AppData\Local\Temp\ipykernel_20608\2342968101.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_salary = round(pd.read_sql_query(query_7_1, connection))


avg
0  84315.0

In [57]:
# проведем анализ основных работодателей в сфере DS
query_7_2 = f''' 
SELECT
    e.name,
    COUNT(v.id)
    
FROM 
    employers e JOIN vacancies v ON e.id = v.employer_id
WHERE 
    v.name ILIKE '%data scientist%'
    OR v.name ILIKE '%data science%'
    OR v.name ILIKE '%исследователь данных%'
    OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%html%') 
    OR v.name ILIKE '%machine learning%'
    OR v.name ILIKE '%машинн%обучен%'
GROUP BY e.name
ORDER BY 2 DESC
LIMIT 7
''' 

In [58]:
# выводим данные по крупнейшим работодателям 
major_ds_employers = pd.read_sql_query(query_7_2, connection)
major_ds_employers

C:\Users\User\AppData\Local\Temp\ipykernel_20608\922133129.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  major_ds_employers = pd.read_sql_query(query_7_2, connection)


name  count
0                   СБЕР     37
1        Bell Integrator     25
2         Банк ВТБ (ПАО)     18
3                     VK     15
4  Positive Technologies     11
5         EvenBet Gaming      9
6                 Яндекс      9

In [59]:
# проведем анализ средней заработной платы по крупнейшим работодателям в ds
query_7_3 = f''' 
SELECT
    e.name,
    COUNT(v.id),
    AVG((COALESCE(salary_from, salary_to) + COALESCE(salary_to, salary_from)) / 2)
FROM 
    employers e JOIN vacancies v ON e.id = v.employer_id
WHERE 
    (
    v.name ILIKE '%data scientist%'
    OR v.name ILIKE '%data science%'
    OR v.name ILIKE '%исследователь данных%'
    OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%html%') 
    OR v.name ILIKE '%machine learning%'
    OR v.name ILIKE '%машинн%обучен%'
    )
AND
    (
    v.salary_from > 0  
    OR v.salary_to > 0
    )
GROUP BY 1
ORDER BY 2 DESC
LIMIT 7
''' 

In [60]:
# выводим результаты анализа по крупнейшим работодателям 
avg_ds_salary_by_employer = round(pd.read_sql_query(query_7_3, connection))
avg_ds_salary_by_employer

C:\Users\User\AppData\Local\Temp\ipykernel_20608\1675359653.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_ds_salary_by_employer = round(pd.read_sql_query(query_7_3, connection))


name  count       avg
0      IntelligiChain      3  233794.0
1            Andersen      3  268863.0
2        НСК ЭкоСфера      2  150000.0
3   DNS Головной офис      2   85000.0
4            Homework      2   90000.0
5           JSA Group      2  300000.0
6  Программный регион      2  127500.0

Общие выводы: 
- База данных содержит сведения о 49 тыс. вакансиях, размещенных 23,5 тыс. работодателями в 1362 локациях; 
- Вакансии представлены почти в 300 сферах деятельности; 
- Данные содержат сведения о вакансиях как в РФ, так и за ее пределами (в основном, в странах ближнего зарубежья); 
- Вилка заработной платы указана почти в половине вакансий (49%), нижняя граница составляет в среднем 71 тыс. рублей, верхняя – 110,5 тыс. рублей; 
- Большая часть вакансий (72%) предусматривает трудоустройство на полный рабочий день; 
- Наибольшее число вакансий представлено в сферах ИТ, Телеком, банковская деятельность, добыча и переработка нефти; 
- Самую большую географию размещения вакансий имеют крупные федеральные компании (Яндекст, Ростелеком), а также менее известные работодатели, работающие в сфере оказания услуг; 
- Разработкой программного обеспечения занимаются порядка 3,5 тыс компаний (порядка 15% от всех работодателей) ;
- Данные содержат сведения о 1771 вакансии в сфере Data Science (~ 3,6% от всех вакансий). Подавляющее большинство вакансий (97%) предусматривает наличие у кандидата опыта работы в данной сфере; 
- Средняя заработная плата в сфере Data Science составляет ~ 177 тыс. рублей, что существенно превышает среднюю заработную плату по всем размещенным вакансиям (84 тыс. рублей); 
- Самое большое число вакансий в сфере Data Science размещено компанией Сбер, работающей в сфере банковской деятельности, на третьей позиции в топе работодателей представлен еще один банк (ВТБ), остальные компании работают в сфере ИТ. Можно сделать вывод, что специалисты DS наиболее востребованы именно в сфере ИТ разработки (сюда же можно отнести и крупнейшие банки, активно развивающие финтех); 
- Крупнейшие работодатели в сфере DS не практикуют указание зарплатной вилки в размещаемых вакансиях, таким образом, анализ уровня предлагаемых зарплат затруднен. 
